### PR TITLE
[AST] A couple of PatternBindingInitializer cleanups

### PR DESCRIPTION
**Pull request by hamishknight**
Remove `SerializedPatternBindingInitializer` & `SerializedDefaultArgumentInitializer` since we can deserialize `PatternBindingInitializer` & `DefaultArgumentInitializer` directly (as I understand it, the `SerializedXXX` decl contexts only exist for cases where we don't want to serialize e.g a full `ClosureExpr`/`TopLevelDecl`, that doesn't apply here). This then allows us to clean up `PatternBindingDecl`'s handling of `PatternBindingInitializer` a bit.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: include/swift/AST/DeclContext.h [hunk 1]
```diff
@@ -115,8 +113,6 @@ enum class DeclContextKind : unsigned {
 /// \see SerializedLocalDeclContext.
```
Comment (slavapestov) [inline include/swift/AST/DeclContext.h:113]: This comment is out of date. Is DeclContextKind::SerializedLocal also dead now?

**Review by slavapestov — APPROVED** [review]
Nice!

**Review by CodaFi — APPROVED** [review]

File: include/swift/AST/DeclContext.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/DeclContext.h:113]: Yeah looks like we can probably just rip out SerializedLocalDeclContext at this point, I'll do that in a follow-up

**Comment by hamishknight** [comment]
Source compat failures are unrelated

File: include/swift/AST/DeclContext.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/DeclContext.h:113]: https://github.com/apple/swift/pull/70983

**Pull request merged**